Add str pointer methods (take 2)

WaffleLapkin opened the pull request:
This PR adds the following functions for pointers to `str`, similar to already existing functions for pointers to `[T]`.

```rust
impl *const str {
    pub const fn len(self) -> usize;
    pub const fn as_ptr(self) -> *const u8;
    pub unsafe fn get_unchecked<I>(self, index: I) -> *const I::Output
    where
        I: SliceIndex<str>;
}

impl *mut str {
    pub const fn len(self) -> usize;
    pub const fn as_mut_ptr(self) -> *mut u8;
    pub unsafe fn get_unchecked_mut<I>(self, index: I) -> *mut I::Output
    where
        I: SliceIndex<str>;
}

impl NonNull<str> {
    pub const fn len(self) -> usize;
    pub const fn as_non_null_ptr(self) -> NonNull<u8>;
    pub const fn as_mut_ptr(self) -> *mut u8;
    pub unsafe fn get_unchecked_mut<I>(self, index: I) -> NonNull<I::Output>
    where
        I: SliceIndex<str>;
}
```

---

This PR also adds `const_str_ptr` and `mut_str_ptr` lang items (they are required for *const str and *mut str impls).

---

This was previously proposed in #85816. This PR includes changes proposed by the review and was created to split these changes from the addition of pointer-to-str-construction methods (#91216).

---

r? @m-ou-se

WaffleLapkin committed:
Add `const_str_ptr` and `mut_str_ptr` lang items

These items allow to make inherent impls for `*const str` and `*mut str`.

WaffleLapkin committed:
Add raw str methods

This patch adds the following methods to `*const str` and `*mut str`:
- `len`
- `as_ptr` (`as_mut_ptr`)
- `get_unchecked` (`get_unchecked_mut`)

Similar methods have already existed for raw slices.

WaffleLapkin committed:
Add nonnull raw str methods

This patch adds the following methods to `NonNull<str>`:
- `len`
- `as_non_null_ptr`
- `as_mut_ptr`
- `get_unchecked_mut`

Similar methods have already existed for raw slices, raw strings and nonnull
raw slices.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4323792932?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d35cd48d63fb302823c4b0c7914a626356858c2b/checks/4323792932/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling memchr v2.4.1
   Compiling std v0.0.0 (/checkout/library/std)
   Compiling compiler_builtins v0.1.52
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error: attributes `#[rustc_const_unstable]` and `#[rustc_const_stable]` require the function or method to be `const`
    |
    |
726 |       #[rustc_const_unstable(feature = "str_ptr_get", issue = "74265")]
    |       ----------------------------------------------------------------- attribute specified here
727 |       #[inline]
728 | /     pub unsafe fn get_unchecked_mut<I>(self, index: I) -> NonNull<I::Output>
729 | |     where
730 | |         I: SliceIndex<str>,
    |
help: make the function or method const
   --> library/core/src/ptr/non_null.rs:728:5
    |
    |
728 | /     pub unsafe fn get_unchecked_mut<I>(self, index: I) -> NonNull<I::Output>
729 | |     where
730 | |         I: SliceIndex<str>,


error: attributes `#[rustc_const_unstable]` and `#[rustc_const_stable]` require the function or method to be `const`
     |
     |
1107 |       #[rustc_const_unstable(feature = "str_ptr_get", issue = "74265")]
     |       ----------------------------------------------------------------- attribute specified here
1108 |       #[inline]
1109 | /     pub unsafe fn get_unchecked<I>(self, index: I) -> *const I::Output
1110 | |     where
1111 | |         I: SliceIndex<str>,
     |
help: make the function or method const
    --> library/core/src/ptr/const_ptr.rs:1109:5
     |
     |
1109 | /     pub unsafe fn get_unchecked<I>(self, index: I) -> *const I::Output
1110 | |     where
1111 | |         I: SliceIndex<str>,


error: attributes `#[rustc_const_unstable]` and `#[rustc_const_stable]` require the function or method to be `const`
     |
     |
1425 |       #[rustc_const_unstable(feature = "str_ptr_get", issue = "74265")]
     |       ----------------------------------------------------------------- attribute specified here
1426 |       #[inline]
1427 | /     pub unsafe fn get_unchecked_mut<I>(self, index: I) -> *mut I::Output
1428 | |     where
1429 | |         I: SliceIndex<str>,
     |
help: make the function or method const
    --> library/core/src/ptr/mut_ptr.rs:1427:5
     |
     |
1427 | /     pub unsafe fn get_unchecked_mut<I>(self, index: I) -> *mut I::Output
1428 | |     where
1429 | |         I: SliceIndex<str>,

error: could not compile `core` due to 3 previous errors
Build completed unsuccessfully in 0:04:17
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #91627) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

m-ou-se commented on library/core/src/ptr/non_null.rs:
```diff
@@ -634,6 +634,107 @@ impl<T> NonNull<[T]> {
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+    ///
+    /// Note that calling this function with an index that does not lie on an UTF-8 sequence boundaries
+    /// is safe, but dereferencing the pointer returned by such call is unsound.
```
Non-UTF-8 `str`s are not UB from a language perspective, it's just that libraries (including std) may assume they are valid UTF-8. So technically dereferencing such a pointer is still sound, even though the resulting value might not uphold any of the invariants that are assumed about the value, potentially causing UB when used afterwards.

WaffleLapkin replied:
Does the following sounds better to you?

> Note that calling this function with an index that does not lie on an UTF-8 sequence boundaries is safe, but dereferencing and using the pointer returned by such call may cause UB.

m-ou-se replied:
It's also that this isn't just about slicing on non-UTF-8 boundaries. The original `*const str` (or `*mut str`) might contain non-UTF-8 somewhere in the middle of the slice. `*const str` and `*mut str` have no guarantees about UTF-8 at all. `&str` and `&mut str` is assumed to be valid UTF-8 though, by many parts of the standard library and many other libraries. Not sure how relevant this warning is on this function, as the `*const str` might already have non-UTF-8 data before slicing it.

m-ou-se commented on library/core/src/ptr/const_ptr.rs:
```diff
@@ -1030,6 +1030,86 @@ impl<T> *const [T] {
+    ///
+    /// The returned value is the number of **bytes**, not the number of characters.
+    ///
+    /// This function is safe, even when the raw string slice cannot be cast to a slice
+    /// reference because the pointer is null or unaligned.
```
The `because` part here seems makes it seem like those are the only cases in which turning the `*const str` into a `&str` would be UB, but there are others too. For example, if the pointer is perfectly valid but the length is too high. Maybe this statement can be simplified. Maybe something that mentions that the pointer/address is irrelevant.

WaffleLapkin replied:
I borrowed this line from `<*const [T]>::len` [documentation](https://doc.rust-lang.org/std/primitive.pointer.html#method.len).

Should I change this in all places to the following:
> This function is safe, even when the raw string slice cannot be safely cast to a slice reference.

Or maybe something else?

m-ou-se replied:
Ah I see. Then we should change it there as well. That edit sounds good. Thank you.

I suppose another option is to add `, for example` after `slice reference`, or add `(e.g. because ..)` in parentheses. Then there's still some examples, without implying that they're the only cases.

m-ou-se reviewed (changes requested):
Thanks for splitting that PR!

This looks great. I only have some small comments on the documentation:

m-ou-se commented on library/core/src/lib.rs:
```diff
@@ -140,6 +140,7 @@
 #![feature(const_default_impls)]
 #![feature(duration_consts_float)]
 #![feature(maybe_uninit_uninit_array)]
+#![cfg_attr(not(bootstrap), feature(const_str_ptr_len))]
```
Could you move this a few lines up so it's sorted correctly between the other const_ features? Thanks!

WaffleLapkin committed:
Adjust documentation for some pointer-to-slice methods

This commit changes documentation of the following methods as proposed
by the PR review
- `<*const [T]>::len`
- `<*mut [T]>::len`
- `<*const str>::len`
- `<*mut str>::len`
- `<*const str>::get_unchecked`
- `<*mut str>::get_unchecked_mut`
- `NonNull::<str>::get_unchecked_mut`

WaffleLapkin commented:
@m-ou-se I've edited the documentation as you suggested

WaffleLapkin commented:
Actually... I now dought that these APIs are useful.

While `&str` and `&[u8]` are different in that `&str` guarantees more about the data, `*const str` and `*const [u8]` are not different at all! As discussed while editing safety comments, it's totally fine for `*const str` to point to non-utf8 bytes, for example.

So I'm not sure that using and operating on `*const str` actually gives us any advantages over using `*const [u8]`.

WaffleLapkin commented:
I'm going to close this PR per the reasoning above. Feel free to ping me, if you disagree and think that these APIs are useful.
